File: src/com/aopphp/go/pointcut/TruePointcut.java
package com.aopphp.go.pointcut;

import com.jetbrains.php.lang.psi.elements.PhpNamedElement;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * Canonical pointcut instance that matches all points.
 */
public class TruePointcut implements Pointcut
{
    private static final Set<KindFilter> KIND_ALL = new HashSet<KindFilter>(Arrays.asList(KindFilter.values())) ;

    /**
     * Filter for class
     */
    protected PointFilter classFilter = TruePointFilter.getInstance();

    private Set<KindFilter> kind = KIND_ALL;

    public TruePointcut() {
    }

    public TruePointcut(Set<KindFilter> kind) {
        this.kind = kind;
    }

    @Override
    public PointFilter getClassFilter() {
        return classFilter;
    }

    public void setClassFilter(PointFilter classFilter) {
        this.classFilter = classFilter;
    }

    @Override
    public boolean matches(PhpNamedElement element) {
        return true;
    }

    @Override
    public Set<KindFilter> getKind() {
        return kind;
    }
}

File: src/main/kotlin/com/aopphp/go/pointcut/TruePointcut.kt
package com.aopphp.go.pointcut

import com.jetbrains.php.lang.psi.elements.PhpNamedElement

open class TruePointcut @JvmOverloads constructor(
    private val _kind: Set<KindFilter> = KindFilter.entries.toSet(),
    @set:JvmName("setClassFilter")
    var _classFilter: PointFilter = TruePointFilter
) : Pointcut {
    override fun getClassFilter() = _classFilter
    override fun matches(element: PhpNamedElement) = true
    override fun getKind() = _kind

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TruePointcut) return false
        return _classFilter == other._classFilter && _kind == other._kind
    }

    override fun hashCode() = 31 * _classFilter.hashCode() + _kind.hashCode()
}
